src/mpid/ch4/shm/posix/posix_coll_gpu_ipc.h

```diff
@@ -234,11 +256,12 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_alltoall_gpu_ipc_read(const void *s
     MPIR_CHKLMEM_MALLOC(reqs, MPL_gpu_request *, sizeof(MPL_gpu_request) * comm_size, mpi_errno,
                         "Memcpy requests", MPL_MEM_COLL);
     for (int i = 0; i < comm_size; i++) {
-        char *temp_recv = (char *) recv_mem_addr + i * data_sz;
-        char *temp_send = (char *) (remote_bufs[i]) + my_rank * data_sz;
+        int target = (my_rank + 1 + i) % comm_size;
+        char *temp_recv = (char *) recv_mem_addr + target * data_sz;
+        char *temp_send = (char *) (remote_bufs[target]) + my_rank * data_sz;
         /* get engine type */
         MPL_gpu_engine_type_t engine_type =
-            MPIDI_IPCI_choose_engine(ipc_handles[i].gpu.global_dev_id,
+            MPIDI_IPCI_choose_engine(ipc_handles[target].gpu.global_dev_id,
```

Comment:
These fixes belong to separate commit, right?

Reply:
Right. Added a separate commit for this optimization.